File: main-svc/src/main/java/com/chilly/main_svc/service/VisitService.java
package com.chilly.main_svc.service;

import com.chilly.main_svc.dto.SaveVisitRequest;
import com.chilly.main_svc.dto.VisitDto;
import com.chilly.main_svc.exception.AccessDeniedException;
import com.chilly.main_svc.exception.NoSuchEntityException;
import com.chilly.main_svc.mapper.VisitDtoMapper;
import com.chilly.main_svc.model.Place;
import com.chilly.main_svc.model.User;
import com.chilly.main_svc.model.Visit;
import com.chilly.main_svc.repository.PlaceRepository;
import com.chilly.main_svc.repository.VisitRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.Date;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class VisitService {

    private final VisitRepository visitRepository;
    private final VisitDtoMapper visitDtoMapper;
    private final PlaceRepository placeRepository;
    private final UserService userService;


    public void saveVisit(Long userId, SaveVisitRequest request) {
        User user = userService.findUserOrException(userId);
        Place place = findPlaceOrException(request.getPlaceId());
        Date date = Optional.ofNullable(request.getTimestamp()).orElseGet(Date::new);

        Visit visit = Visit.builder()
                .place(place)
                .user(user)
                .date(date)
                .build();

        visitRepository.save(visit);
    }

    public List<VisitDto> getVisited(Long userId) {
        User user = userService.findUserOrException(userId);
        return user.getVisits().stream()
                .map(visitDtoMapper::toDto)
                .toList();
    }

    private Place findPlaceOrException(Long id) {
        return placeRepository.findById(id)
                .orElseThrow(() -> new NoSuchEntityException("No place with id =" + id));
    }

    public void deleteVisit(Long userId, Long visitId) {
        Visit visit = visitRepository.findById(visitId)
                .orElseThrow(() -> new NoSuchEntityException("No visit with id =" + visitId));

        if (fromOtherUser(visit, userId)) {
            throw new AccessDeniedException("Cannot delete visit of another user");
        }

        visitRepository.deleteById(visitId);
    }

    private boolean fromOtherUser(Visit visit, Long userId) {
        return !Objects.equals(visit.getUser().getId(), userId);
    }
}

File: places-svc/src/main/kotlin/com/chilly/places_svc/service/VisitService.kt
package com.chilly.places_svc.service

import com.chilly.places_svc.mapper.VisitMapper
import com.chilly.places_svc.model.Visit
import com.chilly.places_svc.repository.PlaceRepository
import com.chilly.places_svc.repository.VisitRepository
import org.chilly.common.dto.SaveVisitRequest
import org.chilly.common.exception.AccessDeniedException
import org.chilly.common.exception.NoSuchEntityException
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import java.util.*

@Service
class VisitService(
    private val visitRepository: VisitRepository,
    private val visitMapper: VisitMapper,
    private val placeRepository: PlaceRepository,
) {

    fun saveVisit(userId: Long, request: SaveVisitRequest) {
        val place = findPlaceOrException(request.placeId)
        val date = request.timestamp ?: Date()

        val visit = Visit(date).apply {
            this.place = place
            this.userId = userId
        }

        visitRepository.save(visit)
    }

    fun getVisited(userId: Long) = visitRepository.findAllByUserId(userId)
        .map(visitMapper::toDto)

    fun deleteVisit(userId: Long, visitId: Long) {
        val visit = findVisitOrException(visitId)

        if (visit.userId != userId) {
            throw AccessDeniedException("cannot remove visit from another user")
        }

        visitRepository.deleteById(visitId)
    }

    private fun findPlaceOrException(id: Long) = placeRepository.findByIdOrNull(id)
        ?: throw NoSuchEntityException("No place with id=$id")

    private fun findVisitOrException(id: Long) = visitRepository.findByIdOrNull(id)
        ?: throw NoSuchEntityException("No visit with id=$id")
}